### DISCUSSION TITLE
Functional Dbs / generating a server with prisma schema file.

### DISCUSSION THREAD

**Discussion by ericbets** [prisma/prisma#3123] [Q&A] [answered]
Thought I'd start a thread about generating a server api for modifying database records, with typical features eg. input validation and rbac. For example, by adding gpc style rpc syntax into the prisma schema file.

I experimented a bit today with generating a full grpc server based on a prisma schema model.
It turns out that grpc is a nice fit for this because you can use interceptors for rbac & input validation using [grpc-host-builder](https://github.com/litichevskiydv/grpc-host-builder).

```console
model Post {
  id        Int     @id @default(autoincrement())
  title     String
  content   String?
  published Boolean @default(false)
  author    User?   @relation(fields:  [authorId], references: [id])
  authorId  Int?
}

//query types
model PostTitles {
  title String  @references (Post:title)
  content String @references (Post:content)
}

rpc getBlogCatalog(RangeRequest) returns (PostTitles) {}
```

Related possibly of interest:
 [Functional Databases](https://github.com/ericbets/erics-designs/blob/master/funcdb.md)  
[Danby](https://github.com/ericbets/danby)

**Comment by ericbets** [comment]
An update with some further thoughts on this topic:
I think we are going to do this using protocol buffers custom options to encode the prisma spec info, and use that to generate prisma schema. It's supported in the current proto v3 format but only documented in the proto v2 docs. Here is the relevant link for anyone else interested in this: [Custom Options](https://developers.google.com/protocol-buffers/docs/proto#customoptions)

**Reply by ryands17** [reply]
Hey @ericbets 👋 
It would be great if you could reference this discussion in a new feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that we can prioritize it :)

**Reply by ericbets** [reply, marked as answer]
[#3231 ](https://github.com/prisma/prisma/issues/3231) Thanks @ryands17!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
